core/src/data_sources/splitter.rs

```diff
@@ -79,37 +123,89 @@ impl Splitter for BaseV0Splitter {
         let re = Regex::new(r"\s+").unwrap();
         let clean = re.replace_all(text, " ");
         let clean = clean.trim();
-
         // Get the embedder and initialize.
         let mut embedder = provider(provider_id).embedder(model_id.to_string());
         embedder.initialize(credentials).await?;
 
         // Encode the clean text.
-        let encoded = embedder.encode(clean).await?;
-
-        // Split the encoded text into chunks of size max_chunk_size.
-        let mut chunks = Vec::new();
-        let mut chunk = Vec::new();
-        let mut chunk_size = 0;
-        for token in encoded {
-            chunk.push(token);
-            chunk_size += 1;
-            if chunk_size >= max_chunk_size {
-                chunks.push(chunk);
-                chunk = Vec::new();
-                chunk_size = 0;
+        let mut encoded = embedder.encode(clean).await?;
+
+        let mut decoded = vec![];
+
+        while encoded.len() > 0 {
+            let mut current_chunk_size = cmp::min(max_chunk_size, encoded.len());
+            let tokenized_chunk = &encoded[0..current_chunk_size];
+
+            match self.decode_chunk_with_remainder(&embedder, tokenized_chunk).await {
+                Ok((chunk_decoded, chunk_remainder)) => {
+                    if chunk_remainder.is_some() {
+                        let chunk_remainder_value = chunk_remainder.unwrap();
+                        if chunk_remainder_value.len() >= max_chunk_size {
+                            return Err(anyhow!("Could not tokenize the provided document.."));
+                        }
+                        current_chunk_size = current_chunk_size - chunk_remainder_value.len();
+                    }
+
+                    if chunk_decoded.is_some() {
+                        decoded.push(chunk_decoded.unwrap());
+                    }
+                }
+                Err(e) => {
+                    return Err(e);
+                }
             }
-        }
-        if chunk.len() > 0 {
-            chunks.push(chunk);
+            encoded = encoded[current_chunk_size..].to_vec();
```

Comment:
last thing that worries me is if we persistently fail to decode at the end and end up in an infinite loop. This should not be possible if the world act as expected but we never know and really don't want to end up in an infinite loop.

I would add a check here that current_chunk_size > 0 if not, error